File: ProjectTimer/ProjectTimer/timer.cpp
#include "timer.h"

using namespace std::chrono;

#pragma region timer Methods
// Constructors and Destructors
timer::timer()
{
	startPoint = steady_clock::now();
	activeStatus = true;
	millisecondsElapsed = 0ns;
	name = "";
}


timer::timer(string _name) : name(_name)
{
	startPoint = steady_clock::now();
	activeStatus = true;
	millisecondsElapsed = 0ns;
}


// Other Public Methods
void timer::pauseTimer()
{
	if (!isTimerActive())
	{
		// Do nothing
	}
	else
	{
		activeStatus = false;
		millisecondsElapsed = millisecondsElapsed + (steady_clock::now() - startPoint);
	}
}


void timer::startTimer()
{
	auto newStartPoint = steady_clock::now();
	if (isTimerActive())
	{
		// Do nothing
	}
	else
	{
		activeStatus = true;
		startPoint = newStartPoint;
	}
}


void timer::resetTimer()
{
	startPoint = steady_clock::now();
	millisecondsElapsed = 0ns;
}


nanoseconds timer::getTimeElapsed()
{
	if (isTimerActive())
	{
		return steady_clock::now() - startPoint + millisecondsElapsed;
	}
	else 
	{
		return millisecondsElapsed;
	}
}


bool timer::isTimerActive()
{
	return activeStatus;
}


void timer::addTime(milliseconds _milliseconds)
{
	millisecondsElapsed = millisecondsElapsed + _milliseconds;
}


void timer::subtractTime(chrono::milliseconds _milliseconds)
{
	millisecondsElapsed = millisecondsElapsed - _milliseconds;
}


void timer::setName(string _name)
{
	name = _name;
}


string timer::getName()
{
	return name;
}


// Private Methods
void timer::setActiveStatus(bool _status)
{
	activeStatus = _status;
}
#pragma endregion


#pragma region totalTimer Methods
// Constructors and Destructors
totalTimer::totalTimer() : timer()
{
	vector<timer> subTimers ( {timer("Enter name of timer")});
	activeSubTimerPointer = &subTimers.front();
}


totalTimer::totalTimer(string _name, vector<timer> _subTimers) : timer(_name)
{
	subTimers = _subTimers;
	activeSubTimerPointer = &subTimers.front();
}


// Other Public Methods
void totalTimer::pauseTimer()
{
	// TODO: improve with threading?
	pauseTimer();
	pauseActiveSubTimer();
}


void totalTimer::addSubTimer(timer _subTimer)
{
	subTimers.push_back(_subTimer);
}


void totalTimer::removeSubTimer(int _position)
{
	subTimers.erase(subTimers.begin() + _position);
}


vector<timer> totalTimer::getSubTimers()
{
	return subTimers;
}


void totalTimer::setAndStartActiveSubTimer(int _position)
{
	activeSubTimerPointer = &subTimers.at(_position);
	startActiveSubTimer();
}


timer totalTimer::getActiveSubTimer()
{
	return *activeSubTimerPointer;
}


// Private Methods
void totalTimer::startActiveSubTimer()
{
	// TODO: improve with threading?
	activeSubTimerPointer->startTimer();
	startTimer();
}


void totalTimer::pauseActiveSubTimer()
{
	activeSubTimerPointer->pauseTimer();
}
#pragma endregion

File: WinFormsApp1/WinFormsApp1/Timer.cs
namespace MyTimers
{
    internal class Timer
    {
        #region Constructors
        /// <summary>
        /// Creates a <c>Timer</c> object.
        /// </summary>
        public Timer()
        {
            _startPoint = DateTime.Now;
            _timeElapsed = TimeSpan.Zero;
            _activeStatus = false;
        }

        /// <summary>
        /// Creates a <c>Timer</c> object with name <paramref name="name"/>.
        /// </summary>
        /// <param name="name">Name of the <c>Timer</c></param>
        public Timer(string name) : this()
        {
            _name = name;
        }
        #endregion


        #region Public Methods
        /// <summary>
        /// Pauses the <c>Timer</c>.
        /// </summary>
        public void PauseTimer()
        {
            var nowDateTime = DateTime.Now;
            if (IsTimerActive())
            {
                _activeStatus = false;
                _timeElapsed += (nowDateTime - _startPoint);
            }
        }

        /// <summary>
        /// Starts the <c>Timer</c>.
        /// </summary>
        public void StartTimer()
        {
            var newStartPoint = DateTime.Now;
            if (!IsTimerActive())
            {
                _activeStatus = true;
                _startPoint = newStartPoint;
            }
        }

        /// <summary>
        /// Resets the <c>Timer</c> to 0.
        /// </summary>
        public void ResetTimer()
        {
            var newStartPoint = DateTime.Now;
            _activeStatus = false;
            
            _startPoint = newStartPoint;
            _timeElapsed = TimeSpan.Zero;           
        }

        /// <summary>
        /// Gets the time elapsed from the <c>Timer</c>.
        /// </summary>
        /// <returns>Time Elapsed <c>TimeSpan</c></returns>
        public TimeSpan GetTimeElapsed()
        {
            var nowDateTime = DateTime.Now;
            if (IsTimerActive())
            {
                return nowDateTime - _startPoint + _timeElapsed;
            }
            else
            {
                return _timeElapsed;
            }
        }

        /// <summary>
        /// Gets the status of the <c>Timer</c>.
        /// </summary>
        /// <returns><c>True</c> if the <c>Timer</c> is active, <c>False</c> if the <c>Timer</c> is inactive.</returns>
        public bool IsTimerActive()
        {
            return _activeStatus;
        }

        /// <summary>
        /// Adds <paramref name="time"/> to the <c>Timer</c>.
        /// </summary>
        /// <param name="time">Time to be added to the timer.</param>
        public void AddTime(TimeSpan time)
        {
            _timeElapsed += time;
        }

        /// <summary>
        /// Subtracts <paramref name="time"/> to the <c>Timer</c>.
        /// </summary>
        /// <param name="time">Time to be subtracted from the timer.</param>
        public void SubtractTime(TimeSpan time)
        {
            TimeSpan newTimeElapsed = new[] { TimeSpan.Zero, _timeElapsed - time }.Max();
            if (newTimeElapsed == TimeSpan.Zero)
            {
                _startPoint = DateTime.Now;
            }
            _timeElapsed = newTimeElapsed;
        }

        /// <summary>
        /// Sets the name of the <c>Timer</c>.
        /// </summary>
        /// <param name="name">The name of the timer.</param>
        public void SetName(string name)
        {
            _name = name;
        }

        /// <summary>
        /// Gets the name of the <c>Timer</c>.
        /// </summary>
        /// <returns>Name of the timer</returns>
        public string GetName()
        {
            return _name;
        }
        #endregion


        #region Private Fields
        private DateTime _startPoint;
        private TimeSpan _timeElapsed;
        private bool _activeStatus;
        private string _name = "";
        #endregion
    }
}
